src/v/kafka/client/direct_consumer/fetcher.cc

```diff
@@ -153,20 +303,28 @@ bool fetcher::maybe_update_fetch_offset(
 
     vlog(
       logger().trace,
-      "[broker: {}] Updating {}/{} fetch offset to {}",
+      "[broker: {}] Updating {}/{} fetch offset from {} to {} {{hwm: {}}}",
       _id,
       topic,
       partition_id,
-      kafka::next_offset(last_received));
+      p_it->second.fetch_offset,
+      kafka::next_offset(last_received),
+      high_watermark);
+    p_it->second.high_watermark = high_watermark;
     p_it->second.fetch_offset = kafka::next_offset(last_received);
+    // we updated the fetch offset, so we should sync to with the broker's
+    // fetch session on the next request
+    p_it->second.incremental_include = true;
```

Comment:
i was under the impression that most of the time the fetch request should be empty. This way we will keep including the partitions into it.

Reply:
my understanding is that when we get new data, that means that the partition's visible offset is above the fetch offset in the session cache (broker side) , so _that_ is when we want to include that partition in the next incremental fetch - i.e. to put the broker's cached fetch offset in sync w/ the hwm. if the hwm hasn't changed, that partition will be omitted from the next response.

basically we include a partition in the request when we want the broker to stop returning data, not when we want it to start. that's my mental model of it anyway, and it does seem to work as advertised (see fixture tests `*Contents*`).